### PR TITLE
Improve locality

### PR DESCRIPTION
**Pull request by aegkmq**
Changing the order of two loops provides ~10% speed up. (`238 tok/s -> 260 tok/s`)

### CODE REVIEW

File: run.c [hunk 1]
```diff
@@ -277,14 +277,16 @@ void transformer(int token, int pos, Config* p, RunState* s, TransformerWeights*
...
-            
+
             // weighted sum of the values, store back into xb
-            for (int i = 0; i < head_size; i++) {
-                float val = 0.0f;
-                for (int t = 0; t <= pos; t++) {
-                    val += att[t] * s->value_cache[loff + t * dim + h * head_size + i]; // note bad locality
+            float* xb = s->xb + h * head_size;
+            memset(xb, 0, head_size * sizeof(float));
+            for (int t = 0; t <= pos; t += 1) {
```
Comment (kroggen) [inline run.c:284]: `t += 1` => `t++`

File: run.c [hunk 2]
```diff
@@ -277,14 +277,16 @@ void transformer(int token, int pos, Config* p, RunState* s, TransformerWeights*
...
+                float* v = s->value_cache + loff + t * dim + h * head_size;
+                float a = s->att[h * p->seq_len + t];
+                for (int i = 0; i < head_size; i += 1) {
```
Comment (kroggen) [inline run.c:287]: `i += 1` => `i++`

File: run.c [hunk 3]
```diff
@@ -277,14 +277,16 @@ void transformer(int token, int pos, Config* p, RunState* s, TransformerWeights*
...
             // weighted sum of the values, store back into xb
-            for (int i = 0; i < head_size; i++) {
-                float val = 0.0f;
-                for (int t = 0; t <= pos; t++) {
-                    val += att[t] * s->value_cache[loff + t * dim + h * head_size + i]; // note bad locality
+            float* xb = s->xb + h * head_size;
+            memset(xb, 0, head_size * sizeof(float));
+            for (int t = 0; t <= pos; t += 1) {
+                float* v = s->value_cache + loff + t * dim + h * head_size;
+                float a = s->att[h * p->seq_len + t];
```
Comment (kroggen) [inline run.c:286]:
```c
float a = att[t]
```

**Comment by karpathy** [comment]
this is nice ty @aegkmq  and @kroggen

**Pull request merged**